AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/DataStoreConfiguration.swift

```diff
@@ -0,0 +1,112 @@
+//
+// Copyright 2018-2020 Amazon.com,
+// Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Amplify
+import Foundation
+
+/// Error Handler function typealias
+public typealias DataStoreErrorHandler = (AmplifyError) -> Void
+
+/// Holds a reference to both the local `Model` and the remote one during a conflict
+/// resolution. Implementations of the `DataStoreConflictHandler` use this to decide
+/// what the outcome of a conflict should be.
+public struct DataStoreConclictData {
+    public let local: Model
+    public let remote: Model
+}
+
+/// Conflict Handler function typealias. The function is used during a conflict that
+/// could not be resolved and requires a decision from the consumer.
+public typealias DataStoreConflictHandler = (DataStoreConclictData, DataStoreConflictHandlerResolver) -> Void
+
+/// Callback for the `DataStoreConflictHandler`.
+public typealias DataStoreConflictHandlerResolver = (DataStoreConflictHandlerResult) -> Void
+
+/// The conflict resolution result enum.
+public enum DataStoreConflictHandlerResult {
+
+    /// Discard the local changes in favor of the remote ones.
+    case discard
+
+    /// Keep the local changes.
+    case keep
+
+    /// Return a new `Model` instance that should used instead of the local and remote changes.
+    case new(Model)
+}
+
+/// The `DataStore` plugin configuration object.
+public struct DataStoreConfiguration {
+
+    /// A callback function called on unhandled errors
+    public let errorHandler: DataStoreErrorHandler
+    
+    /// A callback called when a conflict could not be resolved by the service
+    public let conflictHandler: DataStoreConflictHandler
```

Comment:
This PR doesn't "wire" these into the code, right? We're just defining the attributes and usage of the `DataStoreConfiguration`, in this PR?

I think it looks good, if so.

We should layer these two handlers in some heft docs. One particular thing that might be worth noting the Apple doc is:
1. The conflict handler runs synchronously, and will block (what actions) until complete?

I like the error handler. I like plain callbacks in general. But, why are we using an error handler callback, instead of broadcasting errors to Hub? Or, will we do both? Does the error handler work synchronously, and block other operations, until complete? If so, which ones? Can we document all of this?

Reply:
Thanks for bringing this up, and all valid points.  I've opened an internal ticket against us to complete this before GA.

Reply:
@wooj2 I think the doco could go right into this PR, no? The error events over Hub might be a good follow up task.